nondex-common/src/main/java/edu/illinois/nondex/common/SystematicRandom.java

```diff
@@ -0,0 +1,144 @@
+                        int current = Integer.parseInt(choiceValues[0].toString());
+                        int maximum = Integer.parseInt(choiceValues[1].toString());
+                        boolean shouldExplore = Boolean.parseBoolean(choiceValues[2].toString());
+                        ExplorationEntry currentMaximum = new ExplorationEntry(current, maximum, shouldExplore);
+                        choices.push(currentMaximum);
+                    } else {
+                        Logger.getGlobal().log(Level.SEVERE, "The 3 ExplorationEntry variable were not stored properly");
+                    }
+
+                }
+            }
+        }
+    }
+
+    public int nextInt(final int maximum) {
+        int current;
+        boolean explore;
+        if (replayIndex < choices.size()) {
+            current =  choices.get(replayIndex).getCurrent();
+        } else {
+            current = 0;
+            if (choices.size() > config.start) {
+                explore = true;
+            } else {
+                explore = false;
+            }
+            ExplorationEntry choiceNums = new ExplorationEntry(current, maximum, explore);
+            choices.push(choiceNums);
+        }
+        replayIndex++;
+        return current;
+    }
+
+    public void endRun() {
+        while (!choices.isEmpty()) {
+            ExplorationEntry currentMaximum = choices.pop();
+            int current = currentMaximum.getCurrent();
+            int maximum = currentMaximum.getMaximum();
+            boolean shouldExplore = false;
+            if (current < maximum - 1) {
+                current++;
+                if (choices.size() > config.start) {
+                    shouldExplore = true;
+                }
+                currentMaximum.setCurrent(current);
+                currentMaximum.setShouldExplore(shouldExplore);
+                choices.push(currentMaximum);
+                replayIndex = 0;
+                try (BufferedWriter bufferedWriter = Files.newBufferedWriter(Paths.get(logFileName))) {
+                    for (ExplorationEntry element : choices) {
+                        String lastAndMax = element.getCurrent() + " " + element.getMaximum()
+                            + " " + element.getShouldExplore();
+                        bufferedWriter.write(lastAndMax);
+                        bufferedWriter.newLine();
+                    }
+                } catch (IOException ioe) {
+                    Logger.getGlobal().log(Level.SEVERE,"Could not write systematic.log file" ,ioe);
+                }
+                return;
+            }
+
+            boolean allFalse = true;
+            for (ExplorationEntry ch: choices) {
+                if (ch.getShouldExplore()) {
+                    allFalse = false;
+                }
+            }
+            if (allFalse) {
+                break;
+            }
+        }
+
+        if (Files.exists(Paths.get(logFileName))) {
```

Comment:
I would not delete the log file. It may be good to keep it in for debugging purposes at least and see its contents, even after the exploration is done, to check that everything is at max. This could be an assertion in your test, that after the exploration is finished, on the stack, current is equal to max in all positions where shouldExplore is true.